tasks/FileUploadTask.md

```diff
@@ -1 +1,153 @@
 # File Upload Task
+
+## Objectives
+
+Uploading a large file can be done by creating an upload session which allows the upload of chunks of the file in sequential API requests. This therefore enables the possibility mechanisms such as resuming the upload or cancelling the upload.
+This task aims to provide a fluent and easy to use mechanism for the consumer to control, monitor and cancel the handling of large uploads.
+
+## Requirements
+
+- Accept a Microsoft.Graph.UploadSession instance to be used for the upload which is created and managed by the user and handed over to the task. The task should use the instance to start/resume the upload.
+- Accept a file/stream to be uploaded
+- The task should provide a mechanism of configuring the upload properties i.e. the upload chunk size
+- The task should provide a mechanism of monitoring the upload status i.e the progress and failures through a callback/Delegate
+- Using the RequestContext, the flag for the FileUploadTask can be set.
+- In the event that a chunk fails to upload, the task should retry to re-upload the chunk for a configurable number of times(default of 3) before giving up and throwing an error/exception for the user to handle..
+
+## Performance Considerations
+
+- The FileUploadTask should not unnecessarily create new streams/buffers to read or write to for the purposes of the small chunks. The memory consumed should be as minimal as possible.
+
+## Telemetry considerations
+
+- Using the RequestContext, the flag for the FileUploadTask can be set to enable monitoring of the use of the task.
+
+## Example usage
+
+### CSharp
+
+Create a handler for the progress of the upload
+
+```CSharp
+
+public class ProgressCallback : IProgressCallback
```

Comment:
We should consider using the [AsyncMonitor](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dc8ae2e7cc49a8f2240d2cba707ef52044ecdfa6/src/Microsoft.Graph.Core/Requests/AsyncMonitor.cs) for this. I don't think we use anywhere so it is worth a look.

Reply:
To me, it looks like the AsyncMonitor is better suited for a long running operation running on graph service (e.g. [creating a copy](https://docs.microsoft.com/en-us/graph/long-running-actions-overview?tabs=http) of a large file already on onedrive) and one would be polling the service for the progress at a frequency of their choosing. 
However, we want to provide a feed back mechanism to what the client SDK is doing for the user and provide mechanisms to show progress and react to failures as the client SDK does the upload task.
@MIchaelMainer

Reply:
Makes sense.